# Exploratory Data Analysis - Predict Customer Churn

## 1. Setup and data loading

In [1]:
import csv
import pandas as pd

df = pd.read_csv('train.csv')
print(f"Loaded {len(df)} rows and {len(df.columns)} columns from train.csv")

df.head()

Loaded 594194 rows and 21 columns from train.csv


,id,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,0,Male,0,Yes,Yes,29,Yes,No,DSL,Yes,...,Yes,Yes,No,No,One year,Yes,Mailed check,60.10,1653.85,No
1,1,Male,0,Yes,Yes,58,Yes,No,DSL,Yes,...,No,Yes,Yes,No,Two year,No,Credit card (automatic),69.50,3778.20,No
2,2,Male,0,Yes,No,58,Yes,Yes,Fiber optic,No,...,No,No,Yes,Yes,Month-to-month,Yes,Electronic check,100.40,5841.35,No
3,3,Female,0,No,No,1,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,69.70,70.70,Yes
4,4,Female,0,No,No,1,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.45,70.45,Yes


## 2. Data Overview 

In [7]:
df.shape

(594194, 21)

> The dataset has 594,194 rows with 21 columns

---

In [ ]:
df.isna().sum()

> There are no missing or empty values in the dataset.

---

In [6]:
df.describe(include='all')

,id,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
count,594194.000000,594194,594194.000000,594194,594194,594194.000000,594194,594194,594194,594194,...,594194,594194,594194,594194,594194,594194,594194,594194.000000,594194.000000,594194
unique,NaN,2,NaN,2,2,NaN,2,3,3,3,...,3,3,3,3,3,2,4,NaN,NaN,2
top,NaN,Female,NaN,Yes,No,NaN,Yes,No,Fiber optic,No,...,No,No,Yes,Yes,Month-to-month,Yes,Electronic check,NaN,NaN,No
freq,NaN,298738,NaN,309554,414362,NaN,557893,283384,272386,289474,...,247377,288571,240301,241435,298918,365579,215372,NaN,NaN,460377
mean,297096.500000,NaN,0.114102,NaN,NaN,36.577258,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,65.866223,2494.377057,NaN
std,171529.177263,NaN,0.317936,NaN,NaN,25.061922,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,31.067444,2353.916710,NaN
min,0.000000,NaN,0.000000,NaN,NaN,1.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,18.250000,18.800000,NaN
25%,148548.250000,NaN,0.000000,NaN,NaN,12.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,29.900000,639.650000,NaN
50%,297096.500000,NaN,0.000000,NaN,NaN,35.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,74.100000,1433.650000,NaN
75%,445644.750000,NaN,0.000000,NaN,NaN,62.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,90.800000,4263.800000,NaN


- Churn is imbalanced: 
  * 460,377 "No" out of 594,194 ≈ 77.5% No / 22.5% Yes.
- gender nearly 50/50 (Female top at 298,738) 
  * likely uninformative.
- SeniorCitizen mean 0.114 
  * only ~11% are seniors.
- Partner roughly split (Yes top, 309,554 ≈ 52%).
- Dependents skewed: ~70% have No.
- PhoneService: ~94% Yes (557,893)
  * low variance, weak predictor.
- InternetService: 
  * top is Fiber optic (272,386)
  * 3 categories (DSL / Fiber / No).


In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 594194 entries, 0 to 594193
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   id                594194 non-null  int64  
 1   gender            594194 non-null  str    
 2   SeniorCitizen     594194 non-null  int64  
 3   Partner           594194 non-null  str    
 4   Dependents        594194 non-null  str    
 5   tenure            594194 non-null  int64  
 6   PhoneService      594194 non-null  str    
 7   MultipleLines     594194 non-null  str    
 8   InternetService   594194 non-null  str    
 9   OnlineSecurity    594194 non-null  str    
 10  OnlineBackup      594194 non-null  str    
 11  DeviceProtection  594194 non-null  str    
 12  TechSupport       594194 non-null  str    
 13  StreamingTV       594194 non-null  str    
 14  StreamingMovies   594194 non-null  str    
 15  Contract          594194 non-null  str    
 16  PaperlessBilling  594194 non-nu

## 3. Data Quality Assessment

## 4. Univariate Analysis

## 5. Bivariate Analysis